circuit_breaker.go

```diff
@@ -0,0 +1,161 @@
+	failThreshold, successThreshold int
+
+	state                   atomic.Value // circuitBreakerState
+	failCount, successCount atomic.Uint32
+	lastFail                time.Time
+}
+
+// NewCircuitBreaker creates a new CircuitBreaker with default settings.
+// The default settings are:
+// - Timeout: 10 seconds
+// - FailThreshold: 3
+// - SuccessThreshold: 1
+// - Policies: CircuitBreaker5xxPolicy
+func NewCircuitBreaker() *CircuitBreaker {
+	cb := &CircuitBreaker{
+		policies:         []CircuitBreakerPolicy{CircuitBreaker5xxPolicy},
+		timeout:          10 * time.Second,
+		failThreshold:    3,
+		successThreshold: 1,
+	}
+	cb.state.Store(closed)
+	return cb
+}
+
+// SetPolicies sets the CircuitBreakerPolicy's that the CircuitBreaker will use to determine whether a response is a failure.
+func (cb *CircuitBreaker) SetPolicies(policies []CircuitBreakerPolicy) *CircuitBreaker {
+	cb.policies = policies
+	return cb
+}
+
+// SetTimeout sets the timeout duration for the CircuitBreaker
+func (cb *CircuitBreaker) SetTimeout(timeout time.Duration) *CircuitBreaker {
+	cb.timeout = timeout
+	return cb
+}
+
+// SetFailThreshold sets the number of failures that must occur within the timeout duration for the CircuitBreaker to
+// transition to the Open state.
+func (cb *CircuitBreaker) SetFailThreshold(threshold int) *CircuitBreaker {
+	cb.failThreshold = threshold
+	return cb
+}
+
+// SetSuccessThreshold sets the number of successes that must occur to transition the CircuitBreaker from the Half-Open state
+// to the Closed state.
+func (cb *CircuitBreaker) SetSuccessThreshold(threshold int) *CircuitBreaker {
+	cb.successThreshold = threshold
+	return cb
+}
+
+// CircuitBreakerPolicy is a function that determines whether a response should trip the circuit breaker
+type CircuitBreakerPolicy func(resp *http.Response) bool
+
+// CircuitBreaker5xxPolicy is a CircuitBreakerPolicy that trips the circuit breaker if the response status code is 500 or greater
+func CircuitBreaker5xxPolicy(resp *http.Response) bool {
+	return resp.StatusCode > 499
+}
+
+var ErrCircuitBreakerOpen = errors.New("resty: circuit breaker open")
+
+type circuitBreakerState uint32
+
+const (
+	closed circuitBreakerState = iota
+	open
+	halfOpen
+)
+
+func (cb *CircuitBreaker) getState() circuitBreakerState {
+	return cb.state.Load().(circuitBreakerState)
+}
+
+func (cb *CircuitBreaker) allow() error {
+	if cb == nil {
+		return nil
+	}
+
+	if cb.getState() == open {
+		return ErrCircuitBreakerOpen
+	}
+
+	return nil
+}
+
+func (cb *CircuitBreaker) applyPolicies(resp *http.Response) {
+	if cb == nil {
+		return
+	}
+
+	failed := false
+	for _, policy := range cb.policies {
+		if policy(resp) {
+			failed = true
+			break
+		}
+	}
+
+	if failed {
+		if cb.failCount.Load() > 0 && time.Since(cb.lastFail) > cb.timeout {
+			cb.failCount.Store(0)
+		}
+
+		switch cb.getState() {
+		case closed:
+			cb.failCount.Add(1)
+			if cb.failCount.Load() >= uint32(cb.failThreshold) {
```

Comment:
That's not the way to do

```suggestion
			failCount := cb.failCount.Add(1)
			if failCount >= cb.failThreshold {
```